<h1>Mục lục<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Import-các-thư-viện-cần-thiết" data-toc-modified-id="Import-các-thư-viện-cần-thiết-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Import các thư viện cần thiết</a></span></li><li><span><a href="#Thu-thập-dữ-liệu" data-toc-modified-id="Thu-thập-dữ-liệu-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Thu thập dữ liệu</a></span></li><li><span><a href="#Khám-phá-dữ-liệu" data-toc-modified-id="Khám-phá-dữ-liệu-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Khám phá dữ liệu</a></span><ul class="toc-item"><li><span><a href="#Đọc-dữ-liệu-từ-file-csv-vào-dataframe" data-toc-modified-id="Đọc-dữ-liệu-từ-file-csv-vào-dataframe-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Đọc dữ liệu từ file csv vào dataframe</a></span></li><li><span><a href="#Dữ-liệu-gồm-có-bao-nhiêu-dòng-và-bao-nhiêu-cột?" data-toc-modified-id="Dữ-liệu-gồm-có-bao-nhiêu-dòng-và-bao-nhiêu-cột?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?</a></span></li><li><span><a href="#Mỗi-dòng-có-ý-nghĩa-gì?-Có-vấn-đề-các-dòng-có-ý-nghĩa-khác-nhau-không?" data-toc-modified-id="Mỗi-dòng-có-ý-nghĩa-gì?-Có-vấn-đề-các-dòng-có-ý-nghĩa-khác-nhau-không?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?</a></span></li><li><span><a href="#Dữ-liệu-có-các-dòng-bị-lặp-không?" data-toc-modified-id="Dữ-liệu-có-các-dòng-bị-lặp-không?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Dữ liệu có các dòng bị lặp không?</a></span></li><li><span><a href="#Mỗi-cột-có-ý-nghĩa-gì?" data-toc-modified-id="Mỗi-cột-có-ý-nghĩa-gì?-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Mỗi cột có ý nghĩa gì?</a></span></li><li><span><a href="#Mỗi-cột-hiện-đang-có-kiểu-dữ-liệu-gì?-Có-cột-nào-có-kiểu-dữ-liệu-chưa-phù-hợp-để-có-thể-xử-lý-tiếp-không?" data-toc-modified-id="Mỗi-cột-hiện-đang-có-kiểu-dữ-liệu-gì?-Có-cột-nào-có-kiểu-dữ-liệu-chưa-phù-hợp-để-có-thể-xử-lý-tiếp-không?-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?</a></span></li></ul></li><li><span><a href="#Tiền-xử-lý" data-toc-modified-id="Tiền-xử-lý-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Tiền xử lý</a></span><ul class="toc-item"><li><span><a href="#Chuyển-dtype-của-cột-&quot;date&quot;-sang-datetime." data-toc-modified-id="Chuyển-dtype-của-cột-&quot;date&quot;-sang-datetime.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Chuyển dtype của cột "date" sang datetime.</a></span></li></ul></li><li><span><a href="#Khám-phá-dữ-liệu-(tiếp-tục)" data-toc-modified-id="Khám-phá-dữ-liệu-(tiếp-tục)-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Khám phá dữ liệu (tiếp tục)</a></span><ul class="toc-item"><li><span><a href="#Với-mỗi-cột-có-kiểu-dữ-liệu-dạng-numeric,-các-giá-trị-được-phân-bố-như-thế-nào?" data-toc-modified-id="Với-mỗi-cột-có-kiểu-dữ-liệu-dạng-numeric,-các-giá-trị-được-phân-bố-như-thế-nào?-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?</a></span></li><li><span><a href="#Với-mỗi-cột-có-kiểu-dữ-liệu-dạng-categorical,-các-giá-trị-được-phân-bố-như-thế-nào?" data-toc-modified-id="Với-mỗi-cột-có-kiểu-dữ-liệu-dạng-categorical,-các-giá-trị-được-phân-bố-như-thế-nào?-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?</a></span></li></ul></li><li><span><a href="#Đưa-ra-các-câu-hỏi-cần-trả-lời" data-toc-modified-id="Đưa-ra-các-câu-hỏi-cần-trả-lời-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Đưa ra các câu hỏi cần trả lời</a></span></li><li><span><a href="#Nhìn-lại-quá-trình-làm-đồ-án" data-toc-modified-id="Nhìn-lại-quá-trình-làm-đồ-án-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Nhìn lại quá trình làm đồ án</a></span></li><li><span><a href="#Tài-liệu-tham-khảo" data-toc-modified-id="Tài-liệu-tham-khảo-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Tài liệu tham khảo</a></span></li></ul></div>

## Import các thư viện cần thiết


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Thu thập dữ liệu

Trong bài phân tích này, chúng em sẽ dùng dữ liệu đã được thu thập sẵn là file "fatal-police-shootings-data.csv" đính kèm. Đây dữ liệu của Washington Post chứa hồ sơ về mọi vụ nổ súng gây tử vong ở Hoa Kỳ của các sĩ quan cảnh sát đang thi hành công vụ kể từ ngày 1 tháng 1 năm 2015. Dữ liệu này được lấy từ nguồn GitHub được cập nhật thường xuyên và gốc của dữ liệu là [ở đây](https://github.com/washingtonpost/data-police-shootings).

Theo mô tả trong file "README.md" từ nguồn GitHub được Washington Post public
>In 2015, The Post began tracking more than a dozen details about each killing — including the race of the deceased, the circumstances of the shooting, whether the person was armed and whether the person was experiencing a mental-health crisis — by culling local news reports, law enforcement websites and social media, and by monitoring independent databases such as Killed by Police and Fatal Encounters. The Post conducted additional reporting in many cases.

>The Post is documenting only those shootings in which a police officer, in the line of duty, shoots and kills a civilian — the circumstances that most closely parallel the 2014 killing of Michael Brown in Ferguson, Mo., which began the protest movement culminating in Black Lives Matter and an increased focus on police accountability nationwide. The Post is not tracking deaths of people in police custody, fatal shootings by off-duty officers or non-shooting deaths.
>
>The FBI and the Centers for Disease Control and Prevention log fatal shootings by police, but officials acknowledge that their data is incomplete. Since 2015, The Post has documented more than twice as many fatal shootings by police as recorded on average annually.
>
>The Post’s database is updated regularly as fatal shootings are reported and as facts emerge about individual cases. The Post seeks to make the database as comprehensive as possible. To provide information about fatal police shootings since Jan. 1, 2015, send us an email at policeshootingsfeedback@washpost.com.

File "fatal-police-shootings-data.csv" mà chúng em đính kèm là file dữ liệu chính mà bạn sẽ làm trong bài tập này. Ngoài ra, chúng em còn đính kèm 1 file phụ: file "state.xls" là file cho biết dân số của các bang ở Mỹ vào năm 2019 và nguồn dữ liệu là [ở đây](https://www.ers.usda.gov/data-products/county-level-data-sets/download-data/
)

Cũng theo file "README.md", dữ liệu này được Washington Post public với license như sau:
>The data is published under an Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0) license: https://creativecommons.org/licenses/by-nc-sa/4.0/
>
>About license: You are free to: Share - copy and redistribute the material in any medium or format. Adapt - remix, transform, and build upon the material. The licensor cannot revoke these freedoms as long as you follow the license terms.

## Khám phá dữ liệu 

### Đọc dữ liệu từ file csv vào dataframe

In [2]:
mvt_df = pd.read_csv('data.csv')
mvt_df.head()

id                name        date   manner_of_death       armed   age  \
0   3          Tim Elliot  2015-01-02              shot         gun  53.0   
1   4    Lewis Lee Lembke  2015-01-02              shot         gun  47.0   
2   5  John Paul Quintero  2015-01-03  shot and Tasered     unarmed  23.0   
3   8     Matthew Hoffman  2015-01-04              shot  toy weapon  32.0   
4   9   Michael Rodriguez  2015-01-04              shot    nail gun  39.0   

  gender race           city state  signs_of_mental_illness threat_level  \
0      M    A        Shelton    WA                     True       attack   
1      M    W          Aloha    OR                    False       attack   
2      M    H        Wichita    KS                    False        other   
3      M    W  San Francisco    CA                     True       attack   
4      M    H          Evans    CO                    False       attack   

          flee  body_camera  longitude  latitude  is_geocoding_exact  
0  Not fleeing        False   -123.122    47.247                True  
1  Not fleeing        False   -122.892    45.487                True  
2  Not fleeing        False    -97.281    37.695                True  
3  Not fleeing        False   -122.422    37.763                True  
4  Not fleeing        False   -104.692    40.384                True

### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột? 

In [3]:
mvt_df.shape

(6419, 17)

Dữ liệu gồm 6419 dòng và 17 cột.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Theo file mô tả dữ liệu và theo quan sát sơ bộ về dữ liệu thì một dòng cho biết thông tin chi tiết về  vụ nổ súng gây tử vong đối với một nạn nhân. Có vẻ không có dòng nào bị "lạc loài".

### Dữ liệu có các dòng bị lặp không? 

In [4]:
np.any(mvt_df.duplicated())

False

Không có dòng nào bị lặp.

### Mỗi cột có ý nghĩa gì?

Theo file mô tả dữ liệu thì:
- `id`: a unique identifier for each victim

- `name`: the name of the victim

- `date`: the date of the fatal shooting in YYYY-MM-DD format

- `manner_of_death`:

    - `shot`
    - `shot and Tasered`
- `armed`: indicates that the victim was armed with some sort of implement that a police officer believed could inflict harm

    - `undetermined`: it is not known whether or not the victim had a weapon
    - `unknown`: the victim was armed, but it is not known what the object was
    - `unarmed`: the victim was not armed
- `age`: the age of the victim

- `gender`: the gender of the victim. The Post identifies victims by the gender they identify with if reports indicate that it differs from their biological sex.

    - `M`: Male
    - `F`: Female
    - `None`: unknown
- `race`:

    - `W`: White, non-Hispanic
    - `B`: Black, non-Hispanic
    - `A`: Asian
    - `N`: Native American
    - `H`: Hispanic
    - `O`: Other
    - `None`: unknown
- `city`: the municipality where the fatal shooting took place. Note that in some cases this field may contain a county name if a more specific municipality is unavailable or unknown.

- `state`: two-letter postal code abbreviation

- `signs of mental illness`: News reports have indicated the victim had a history of mental health issues, expressed suicidal intentions or was experiencing mental distress at the time of the shooting.

- `threat_level`: The threat_level column was used to flag incidents for the story by Amy Brittain in October 2015. http://www.washingtonpost.com/sf/investigative/2015/10/24/on-duty-under-fire/ As described in the story, the general criteria for the attack label was that there was the most direct and immediate threat to life. That would include incidents where officers or others were shot at, threatened with a gun, attacked with other weapons or physical force, etc. The attack category is meant to flag the highest level of threat. The other and undetermined categories represent all remaining cases. Other includes many incidents where officers or others faced significant threats.

- `flee`: News reports have indicated the victim was moving away from officers

    - `Foot`
    - `Car`
    - `Not fleeing`
    
The threat column and the fleeing column are not necessarily related. For example, there is an incident in which the suspect is fleeing and at the same time turns to fire at gun at the officer. Also, attacks represent a status immediately before fatal shots by police while fleeing could begin slightly earlier and involve a chase.

- `body_camera`: News reports have indicated an officer was wearing a body camera and it may have recorded some portion of the incident.

- `latitude and longitude`: the location of the shooting expressed as WGS84 coordinates, geocoded from addresses. The coordinates are rounded to 3 decimal places, meaning they have a precision of about 80-100 meters within the contiguous U.S.

- `is_geocoding_exact`: reflects the accuracy of the coordinates. true means that the coordinates are for the location of the shooting (within approximately 100 meters), while false means that coordinates are for the centroid of a larger region, such as the city or county where the shooting happened.

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không? 

In [5]:
mvt_df.dtypes

id                           int64
name                        object
date                        object
manner_of_death             object
armed                       object
age                        float64
gender                      object
race                        object
city                        object
state                       object
signs_of_mental_illness       bool
threat_level                object
flee                        object
body_camera                   bool
longitude                  float64
latitude                   float64
is_geocoding_exact            bool
dtype: object

- Dữ liệu có 17 thuộc tính.
- Nhìn chung hầu hết các thuộc tính đều có kiểu dữ liệu phù hợp. Riêng cột `date` có kiểu dữ liệu `object `chưa phù hợp.

**Vấn đề cần tiền xử lý:**
Do Cột `date` đang có dtype là `object`. Để có thể tiếp tục khám phá thêm về cột này, chúng em sẽ thực hiện bước tiền xử lý là chuyển sang dạng `datetime`.

## Tiền xử lý

### Chuyển dtype của cột "date" sang datetime.

In [6]:
mvt_df['date'] = pd.to_datetime(mvt_df['date'], format="%Y-%m-%d")
print(mvt_df['date'].dtype)

datetime64[ns]


## Khám phá dữ liệu (tiếp tục)

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào? 

`mvt_df` có 4 cột numeric là: "date", "age", "longitude", "latitude". 

Với mỗi cột numeric nhóm sẽ tính tỉ lệ % giá trị thiếu (từ 0 đến 100), min, max. Lưu kết quả vào dataframe `nume_col_profiles_df`. Dataframe này có 3 dòng là "missing_ratio", "min", "max"; và có 4 cột numeric.  

In [7]:
# YOUR CODE HERE
numeric_arr = ["date", "age", "longitude", "latitude"]

def missing_ratio(s):
    return s.isna().mean() * 100
nume_col_profiles_df = mvt_df[numeric_arr].agg([missing_ratio, min, max])
nume_col_profiles_df

date        age   longitude   latitude
missing_ratio                  0.0   4.548995    4.844991   4.844991
min            2015-01-02 00:00:00   6.000000 -158.137000  19.498000
max            2021-06-30 00:00:00  91.000000  -67.867000  71.301000

Missing_ratio và max của các cột có vẻ không có gì bất thường ngoại trừ min, với độ tuổi thấp nhất là 6. Điều này cho thấy 

### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào? 

`mvt_df` có 13 cột categorical là: "id", "name", "manner_of_death", "armed", "gender", "race", "city", "state" 
          , "signs_of_mental_illness", "threat_level", "flee", "body_camera", "is_geocoding_exact".

Với mỗi cột categorical, nhóm tính tỉ lệ % giá trị thiếu (từ 0 đến 100), số lượng giá trị khác nhau, list/array các giá trị khác nhau. Lưu kết quả vào dataframe `cate_col_profiles_df`. Dataframe này có 3 dòng là "missing_ratio", "num_diff_vals", "diff_vals"; và có 13 cột categorical.

In [8]:
# YOUR CODE HERE
categorical_arr = ["id", "name", "manner_of_death", "armed", "gender", "race", "city","state" 
                   ,"signs_of_mental_illness", "threat_level", "flee", "body_camera", "is_geocoding_exact"]
def num_diff_vals(s):    #số lượng giá trị khác nhau
    return pd.Series.nunique(s, False) 

def diff_vals(s):       #list/array các giá trị khác nhau
    return pd.Series.unique(s)

cate_col_profiles_df = mvt_df[categorical_arr].agg([missing_ratio, num_diff_vals, diff_vals])
cate_col_profiles_df

id  \
missing_ratio                                                0.0   
num_diff_vals                                               6419   
diff_vals      [3, 4, 5, 8, 9, 11, 13, 15, 16, 17, 19, 21, 22...   

                                                            name  \
missing_ratio                                           3.754479   
num_diff_vals                                               6160   
diff_vals      [Tim Elliot, Lewis Lee Lembke, John Paul Quint...   

                        manner_of_death  \
missing_ratio                       0.0   
num_diff_vals                         2   
diff_vals      [shot, shot and Tasered]   

                                                           armed       gender  \
missing_ratio                                            3.24038     0.015579   
num_diff_vals                                                 99            3   
diff_vals      [gun, unarmed, toy weapon, nail gun, knife, na...  [M, F, nan]   

                                  race  \
missing_ratio                10.796074   
num_diff_vals                        7   
diff_vals      [A, W, H, B, O, nan, N]   

                                                            city  \
missing_ratio                                                0.0   
num_diff_vals                                               2778   
diff_vals      [Shelton, Aloha, Wichita, San Francisco, Evans...   

                                                           state  \
missing_ratio                                                0.0   
num_diff_vals                                                 51   
diff_vals      [WA, OR, KS, CA, CO, OK, AZ, IA, PA, TX, OH, L...   

              signs_of_mental_illness                   threat_level  \
missing_ratio                     0.0                            0.0   
num_diff_vals                       2                              3   
diff_vals               [True, False]  [attack, other, undetermined]   

                                               flee    body_camera  \
missing_ratio                              6.761178            0.0   
num_diff_vals                                     5              2   
diff_vals      [Not fleeing, Car, Foot, Other, nan]  [False, True]   

              is_geocoding_exact  
missing_ratio                0.0  
num_diff_vals                  2  
diff_vals          [True, False]

## Đưa ra các câu hỏi cần trả lời

## Nhìn lại quá trình làm đồ án

- `Khó khăn`
    
- `Những điều hữu ích học được`
    - Kĩ năng làm việc nhóm.
    - Kĩ năng sử dụng các công cụ hỗ trợ làm đồ án (như github, trello,...).
    - Các kĩ năng khám phá dữ liệu.
    - Hiểu sâu sắc hơn quy trình Khoa học dữ liệu qua việc tự tìm hiểu và làm đồ án.
    
- `Những dự định nếu có thời gian thêm`

## Tài liệu tham khảo

- https://www.washingtonpost.com/graphics/investigations/police-shootings-database/
- https://github.com/washingtonpost/data-police-shootings
- https://www.ers.usda.gov/data-products/county-level-data-sets/download-data/
- https://en.wikipedia.org/wiki/Race_and_ethnicity_in_the_United_States